### PR TITLE
[FEATURE] Add TER release via GitHub actions

### PR DESCRIPTION
**Pull request by lukaszuznanski**
This PR is the resurrection of #113 (after the switch from master to main).

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+          path: ${{ env.TYPO3_EXTENSION_KEY }}
+
+      - uses: shivammathur/setup-php@v2
+        with:
+          php-version: ${{ matrix.php-versions }}
+          extensions: intl, mbstring, xml, soap, zip, curl
+
+      - name: Install TYPO3 TER client
+
+        run: composer global require namelesscoder/typo3-repository-client
```
Comment (oliverklee) [inline .github/workflows/release.yml:39]: As discussed in https://github.com/TYPO3-Documentation/tea/pull/113#discussion_r500963090, we should switch to https://packagist.org/packages/helhum/ter-client.

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+          "^2.0" --prefer-dist --no-progress --no-suggest
+
+      - name: Remove .git folder
+        run: rm -rf ./$TYPO3_EXTENSION_KEY/.git
```
Comment (oliverklee) [inline .github/workflows/release.yml:43]: We should remove more - including the tests, development configuration files etc. (Helmut's TER client automatically excludes any dot files: https://github.com/helhum/ter-client/blob/master/src/ExtensionUploadPacker.php#L235)

Please feel free to steal from one of my extensions and add anything development-only that only tea has:
https://github.com/oliverklee/ext-oelib/blob/master/.travis.yml#L99

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        with:
+          path: ${{ env.TYPO3_EXTENSION_KEY }}
+
+      - uses: shivammathur/setup-php@v2
+        with:
+          php-version: ${{ matrix.php-versions }}
+          extensions: intl, mbstring, xml, soap, zip, curl
+
+      - name: Install TYPO3 TER client
+
```
Comment (oliverklee) [inline .github/workflows/release.yml:38]: I think not having this blank lines would make it easier to see the blocks.

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        run: composer global require namelesscoder/typo3-repository-client
+          "^2.0" --prefer-dist --no-progress --no-suggest
+
+      - name: Remove .git folder
+        run: rm -rf ./$TYPO3_EXTENSION_KEY/.git
+
+      - name: List extension folder
+        run: ls -liAsh ./$TYPO3_EXTENSION_KEY
```
Comment (oliverklee) [inline .github/workflows/release.yml:46]: Shouldn't the current directory be the extension directory (i.e., the Git repository root) by default instead?

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+      - name: Upload EXT:${{ env.TYPO3_EXTENSION_KEY }}
```
Comment (oliverklee) [inline .github/workflows/release.yml:48]: Is there any way we could read this from the composer.json (where the key already should be listed)?

**Review by oliverklee — CHANGES_REQUESTED** [review]
I've carried over my comments from #113.

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+          "^0.1" --prefer-dist --no-progress
+
+      - name: Remove files
+        run: rm -rf CODE_OF_CONDUCT.md Tests codeception.yml phpcs.xml .github .gitlab .editor .Build .gitattributes .editorconfig .gitignore .php_cs.php Resources/Private/.eslintignore Resources/Private/.eslintrc.json Resources/Private/.prettierrc.js Resources/Private/package.json Resources/Private/stylelint.config.js
+
+      - name: List extension folder
+        run: ls -liAsh ./
+
+      - name: Upload EXT:${{ env.TYPO3_EXTENSION_KEY }}
+          as ${{ github.event.ref }} to TER
```
Comment (oliverklee) [inline .github/workflows/release.yml:47]: I think we need to use the `>` operator for multi-line commands (or put it all in once line), e.g., like this:

https://github.com/oliverklee/ext-phpunit/blob/main/.github/workflows/ci.yml#L120

(I'm not completely sure that this is necessary, though).

File: .github/workflows/release.yml [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        run: ter-client upload $TYPO3_EXTENSION_KEY "./$TYPO3_EXTENSION_KEY"
+          -u $TYPO3_USER -p $TYPO3_PASSWORD"
+          -m "${{ github.event.head_commit.message }}"
```
Comment (oliverklee) [inline .github/workflows/release.yml:50]: Currently, we're using the commit message of the last commit that was tagged. Instead, we should use the commit message of the (annotated) tag.

File: .github/workflows/release.yml [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+      - name: Remove files
+        run: rm -rf CODE_OF_CONDUCT.md Tests codeception.yml phpcs.xml .github .gitlab .editor .Build .gitattributes .editorconfig .gitignore .php_cs.php Resources/Private/.eslintignore Resources/Private/.eslintrc.json Resources/Private/.prettierrc.js Resources/Private/package.json Resources/Private/stylelint.config.js
+
+      - name: List extension folder
+        run: ls -liAsh ./
+
+      - name: Upload EXT:${{ env.TYPO3_EXTENSION_KEY }}
+          as ${{ github.event.ref }} to TER
+        run: ter-client upload $TYPO3_EXTENSION_KEY "./$TYPO3_EXTENSION_KEY"
+          -u $TYPO3_USER -p $TYPO3_PASSWORD"
```
Comment (oliverklee) [inline .github/workflows/release.yml:49]: There's an uneven number of double quotes here.

**Review by oliverklee — CHANGES_REQUESTED** [review]

File: .github/workflows/release.yml (hunk 8 above)
Comment (lukaszuznanski) [reply .github/workflows/release.yml:49]: Yea it's still a draft, will work on it in the following days

File: .github/workflows/release.yml [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+          path: ${{ env.TYPO3_EXTENSION_KEY }}
+
+      - uses: shivammathur/setup-php@v2
+        with:
+          php-version: ${{ matrix.php-versions }}
+          extensions: intl, mbstring, xml, soap, zip, curl
+
+      - name: Install TYPO3 TER client
+        run: composer global require helhum/ter-client
+          "^0.1" --prefer-dist --no-progress
```
Comment (oliverklee) [inline .github/workflows/release.yml:38]: As version 0.1.1 has been released, we can use `^0.1.1` now.

File: .github/workflows/release.yml (hunk 7 above)
Comment (lukaszuznanski) [reply .github/workflows/release.yml:65]: Unfortunately, I can't find any reasonable way to read it from release description, do you have any idea how to get this data?

File: .github/workflows/release.yml [hunk 10]
```diff
@@ -0,0 +1,65 @@
+---
+name: "TER deployment"
+
+on:
+  push:
+    tags:
+      - 'v[0-9]+.[0-9]+.[0-9]+'
```
Comment (oliverklee) [inline .github/workflows/release.yml:7]: By the way, would it make sense to use the [release event](https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#release) instead?

File: .github/workflows/release.yml (hunk 7 above)
Comment (oliverklee) [reply .github/workflows/release.yml:64]: Maybe we can take the [release event payload](https://docs.github.com/en/free-pro-team@latest/rest/reference/repos#get-a-release) and use the first line of the release description?

**Comment by oliverklee** [comment]
With tailor: https://github.com/o-ba/tailor_ext/blob/main/.github/workflows/publish.yml

File: .github/workflows/release.yml [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+            echo ::set-output name=comment::Released version ${{ steps.get-version.outputs.version }} of ${{ steps.split.outputs._1 }}
+          else
+            echo ::set-output name=comment::$comment
+          fi
+
+      - name: Setup PHP
+        uses: shivammathur/setup-php@v2
+        with:
+          php-version: 7.4
+          extensions: intl, mbstring, json, zip, curl
```
Comment (oliverklee) [inline .github/workflows/release.yml:75]: Please let's set the Composer version (2) explicitly (like [in the CI workflow](https://github.com/TYPO3-Documentation/tea/blob/main/.github/workflows/ci.yml#L42)).

File: .github/workflows/release.yml [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+  ter-release:
+    needs: set-matrix
+    name: "TYPO3 TER release"
+
+    if: startsWith(github.ref, 'refs/tags/')
+
+    runs-on: ubuntu-20.04
+    strategy:
+      matrix:
+        php-versions: [ 7.4 ]
```
Comment (oliverklee) [inline .github/workflows/release.yml:29]: We're hardcoding the the PHP version in line 74. So there is no need to have the PHP version in a matrix.

File: .github/workflows/release.yml [hunk 13]
```diff
@@ -0,0 +1,81 @@
+---
+name: "TER deployment"
```
Comment (oliverklee) [inline .github/workflows/release.yml:2]: Please let's synchronize the name of the workflow and the file name. I propose to use "publish" for both so we are close to https://github.com/o-ba/tailor_ext/blob/main/.github/workflows/publish.yml.

File: .github/workflows/release.yml [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+jobs:
+  set-matrix:
+    runs-on: ubuntu-20.04
+    outputs:
+      matrix: ${{ steps.set-matrix.outputs.matrix }}
+    steps:
+      - id: set-matrix
+        run: |
+          JSON=$(cat ./composer.json)
+          echo "::set-output name=${JSON//'%'/'%25'}"
```
Comment (oliverklee) [inline .github/workflows/release.yml:17]: What does this job do? (I don't understand it yet …)

File: .github/workflows/release.yml [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+
+on:
+  release:
+    types: [published]
+
+jobs:
+  set-matrix:
+    runs-on: ubuntu-20.04
+    outputs:
+      matrix: ${{ steps.set-matrix.outputs.matrix }}
```
Comment (oliverklee) [inline .github/workflows/release.yml:12]: Should the output really be a matrix, or should/can it be an env var instead?

File: .github/workflows/release.yml [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+        id: split
+        with:
+          msg: ${{ matrix.composer-data.name}}
+          separator: '/'
+
+      - name: Remove files
+        run: >
+          rm -rf CODE_OF_CONDUCT.md Tests codeception.yml phpcs.xml .github .gitlab .editor .Build .gitattributes
+          .editorconfig .gitignore .php_cs.php Resources/Private/.eslintignore Resources/Private/.eslintrc.json
+          Resources/Private/.prettierrc.js Resources/Private/package.json Resources/Private/stylelint.config.js"
```
Comment (oliverklee) [inline .github/workflows/release.yml:59]: What's the trailing double quote doing here?

File: .github/workflows/release.yml [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+
+      - uses: jungwinter/split@v1
+        id: split
+        with:
+          msg: ${{ matrix.composer-data.name}}
+          separator: '/'
+
+      - name: Remove files
+        run: >
+          rm -rf CODE_OF_CONDUCT.md Tests codeception.yml phpcs.xml .github .gitlab .editor .Build .gitattributes
```
Comment (oliverklee) [inline .github/workflows/release.yml:57]: With the [exclude list included in the latest tailor version](https://github.com/TYPO3/tailor/blob/main/conf/ExcludeFromPackaging.php), we only need to remove the following file (due to a [typo in the exclude list](https://github.com/TYPO3/tailor/pull/19)):
```
.php_cs.php
```

File: .github/workflows/release.yml [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+    steps:
+      - id: set-matrix
+        run: |
+          JSON=$(cat ./composer.json)
+          echo "::set-output name=${JSON//'%'/'%25'}"
+
+  # The actual TER upload job
+  ter-release:
+    needs: set-matrix
+    name: "TYPO3 TER release"
```
Comment (oliverklee) [inline .github/workflows/release.yml:22]: Please let's have the name line directly below the key (above the `needs` line) so both are together (as they say the same).

File: .github/workflows/release.yml [hunk 19]
```diff
@@ -0,0 +1,81 @@
+---
+name: "TER deployment"
+
+on:
+  release:
+    types: [published]
+
+jobs:
+  set-matrix:
```
Comment (oliverklee) [inline .github/workflows/release.yml:9]: Please let's give this job a name, too.

**Review by oliverklee — CHANGES_REQUESTED** [review]

File: .github/workflows/release.yml (hunk 14 above)
Comment (lukaszuznanski) [reply .github/workflows/release.yml]: This job get ext version from composer.json. In order to process json file we need to set it to matrix.

File: .github/workflows/release.yml (hunk 15 above)
Comment (lukaszuznanski) [reply .github/workflows/release.yml]: As I commented above

**Comment by lukaszuznanski** [comment]
Solved by: #189

**Pull request closed without merging**